tests/scripts/all.sh

```diff
@@ -4385,14 +4385,14 @@ component_build_no_ssl_srv () {
     msg "build: full config except SSL server, make, gcc" # ~ 30s
     scripts/config.py full
     scripts/config.py unset MBEDTLS_SSL_SRV_C
-    make CC=gcc CFLAGS='-Werror -Wall -Wextra -O1'
+    make CC=gcc CFLAGS='-Werror -Wall -Wextra -O1 -Wmissing-prototypes'
 }
```

Comment:
We're currently passing `-Wmissing-prototypes` when building with CMake, but only inside `library`. Please move the corresponding clauses from `library/CMakeLists.txt` to the toplevel cmakelists.

Reply:
Just to be sure, do you only refer to the `-Wmissing-prototypes` clauses within these lines:

- From GCC:
  https://github.com/Mbed-TLS/mbedtls/blob/98ffc8e7c30a8a3aa6d688e416d299567b4deb9c/library/CMakeLists.txt#L200
- From Clang:
  https://github.com/Mbed-TLS/mbedtls/blob/98ffc8e7c30a8a3aa6d688e416d299567b4deb9c/library/CMakeLists.txt#L204

...to be moved to these lines:

- To GCC:
  https://github.com/Mbed-TLS/mbedtls/blob/98ffc8e7c30a8a3aa6d688e416d299567b4deb9c/CMakeLists.txt#L199
- To Clang:
  https://github.com/Mbed-TLS/mbedtls/blob/98ffc8e7c30a8a3aa6d688e416d299567b4deb9c/CMakeLists.txt#L233

...or all / other specific clauses from the original lines?

And are the target lines correct?

Reply:
Just moving `-Wmissing-prototypes`, not the other `-W`. I think the target lines are correct, or maybe the GCC line needs to be in a version-based conditional, I don't know when `-Wmissing-prototypes` was introduces.

Reply:
Ok, thanks.

Meanwhile I've dug a bit and the oldest reference I could find in GCC's repo is from 1999:
https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=fe50c0eb3bcd01019e3b8e610b3906510a5d4b8e

That's from version `2.96 19991014`:
https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/version.c;h=2c4f8fbc6ca55f810c8c8e306cee5895e9ae3eae;hb=fe50c0eb3bcd01019e3b8e610b3906510a5d4b8e

So I guess given the age and that it worked in `library` all the time, we can safely assume that it will work.

Reply:
Done in d74b452

I've just seen that the previous CI job failed with:
```
tests/suites/test_suite_ctr_drbg.function:99:14: error: unused function 'thread_random_function' [-Werror,-Wunused-function]
```

Guess I have to figure out a way to build exactly the same locally.

Reply:
So I've managed to reproduce the CI failure by calling what I scrapped from the logs:

`tests/scripts/all.sh --seed 4 --keep-going test_cmake_out_of_source`

Before I ran into the same issue as the CI did, a few more errors were raised on each run:

```
tests/src/drivers/platform_builtin_keys.c:59:14: error: no previous prototype for function 'mbedtls_psa_platform_get_builtin_key'
```

And after making the former function `static`, two `unused` errors occurred:
```
tests/src/drivers/platform_builtin_keys.c:59:21: error: unused function 'mbedtls_psa_platform_get_builtin_key'
tests/src/drivers/platform_builtin_keys.c:26:52: error: unused variable 'builtin_keys'
```

When temporarily commenting those out, I finally ran into the same error as the CI did:

```
tests/suites/test_suite_ctr_drbg.function:99:14: error: unused function 'thread_random_function'
```

Which is actually not true, because `grep` revealed that `thread_random_function` is used within the same file at line 402. It's apparently hard to figure it out for the compiler because it is being used in the `TEST_EQUAL` macro:

https://github.com/Mbed-TLS/mbedtls/blob/7777cb7efc29535e84bf54fae32fe4fb8f74ba23/tests/suites/test_suite_ctr_drbg.function#L399-L404

My suggestion would be to make `thread_random_function` a regular function again and add its prototype directly in the line 99 above the definition:

https://github.com/Mbed-TLS/mbedtls/blob/7777cb7efc29535e84bf54fae32fe4fb8f74ba23/tests/suites/test_suite_ctr_drbg.function#L99

The build and all tests successfully passed, once I did this locally and deleted the unused file `tests/src/drivers/platform_builtin_keys.c` as well :)

Shall I proceed as proposed?

I'm not quite sure because of possible implications of `mbedtls_psa_platform_get_builtin_key` being referenced here when `MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS ` is enabled (which is disabled by default):

https://github.com/Mbed-TLS/mbedtls/blob/7777cb7efc29535e84bf54fae32fe4fb8f74ba23/tests/src/psa_test_wrappers.c#L29-L39

https://github.com/Mbed-TLS/mbedtls/blob/7777cb7efc29535e84bf54fae32fe4fb8f74ba23/include/mbedtls/mbedtls_config.h#L1325-L1341

Reply:
`thread_random_function` is defined unconditionally but only used under certain conditions: `TEST_CASE depends_on:MBEDTLS_THREADING_PTHREAD:!MBEDTLS_CTR_DRBG_USE_128_BIT_KEY:!MBEDTLS_AES_ONLY_128_BIT_KEY_LENGTH` means
```
#if defined(MBEDTLS_THREADING_PTHREAD) && !defined(MBEDTLS_CTR_DRBG_USE_128_BIT_KEY) && !defined(MBEDTLS_AES_ONLY_128_BIT_KEY_LENGTH)
```

(I'm not sure why it's 128-bit-AES only.) Either the function needs to remain non-static and to have a prototype, or its definition needs to have a matching guard.

As for `mbedtls_psa_platform_get_builtin_key`, it's used when building tests in configurations with `MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS` enabled, such as the configuration obtained with `scripts/config.py full`.

Reply:
Thanks for the clarification, done in 815f555

Now the build should pass and we are ready for review.